### ISSUE TITLE
Make Telnet an external library

### ISSUE THREAD

**Issue by pekkaklarck** [robotframework/robotframework#3537]
# tl;dr; summary

> The Telnet library should be made an external library. Once that is done, the standard library Telnet can be removed.

# Background

When Robot Framework was developed at Nokia Networks, having a library for communicating with telnet servers was very valuable when testing different network elements. Back then installing Python packages was pain (`pip` or even `easy_install` didn't exist) so it made sense to include the Telnet library as one of the Robot Framework standard libraries. Initially it didn't have any external dependencies either (Python's standard `telnetlib` was enough).

# Current situation

Nowadays installing Python packages with `pip` is easy and libraries don't benefit much from being standard libraries. It is actually a problem as it means fixes need to wait for Robot Framework releases and that major changes can only occur with major Robot Framework versions which are pretty rare. External libraries can have their own release schedules.

The Telnet library now also has an optional (but very useful) dependency to the `pyte` module, which needs to be installed separately anyway. If the Telnet library was an external library, it could have `pyte` as a dependency and installing the whole thing wouldn't be any more complicated as it is now.

A major problem is that I personally don't have much knowledge about telnet in general and thus I'm not a good person to fix problems with the Telnet library or even to review pull requests. Because of that there are several open issues and PRs here related to it.

# `telnetlib` removal

Python 3.13 removed the standard `telnetlib` module that Telnet uses internally. It exists as a [telnetlib-313-and-up](https://pypi.org/project/telnetlib-313-and-up/) module, but that needs to be installed separately. Python 3.13+ users thus need to anyway install a dependency to get Telnet working and installing the whole library wouldn't be any more complicated.

# Proposal

We should make the Telnet library an external library. In practice that means creating a new project at [MarketSquare](https://github.com/marketsquare) and copying related code and tests there. The name of the new library should probably be `TelnetLibrary`, not `Telnet`, to make it possible to use that and not the current standard library.

I can create the new project and release 1.0 version with the current code. I don't have much interest (or knowledge) maintaining the project after that, so we also need to start looking for interested maintainers. Comment this issue if you are one of them!

Once we have the new project in place and 1.0 version released, we can deprecate the current Telnet library and then later remove it.

# Update

Telnet is deprecated in RF 7.5 (#5786) and will be removed in RF 8.0. The core team can help with creating an external TelnetLibrary project, but is not going to do further maintenance.

**Comment by pekkaklarck** [comment]
Issues and PRs above are related to Telnet. Matching issues should be opened to the new project and they can link to these ones.

**Comment by macpijan** [comment]
Hi @pekkaklarck , is it still the case?

We (3mdeb) are users of Telnet library, and we are interested that it is still around and operational.

We are also interested in some small extensions, such as [this one](https://github.com/robotframework/robotframework/pull/2961#issuecomment-422187187), so we do not maintain such patch in fork.

I see there is really no traffic around this library. We could step up as a maintainer for it, as we are using it on a daily basis and are interested to making it compliant with RF releases, or providing small features/fixes, if any. Although we've been using it for years, with very little problems, to be honest.

**Comment by pekkaklarck** [comment]
Thanks for reaching out @macpijan. This is still very much the case and due to `telnetlib` being deprecated (see [PEP 594](https://peps.python.org/pep-0594/)) getting even more important.

It would be awesome if you could take the maintenance responsibility. As you commented, the library works pretty well and maintaining it is likely not too big a task. There is, however, some initial work getting the new project started, but I can help if needed.

How should we get started? We can discuss here or on our Slack. As I commented earlier, [MarketSquare](https://github.com/marketsquare) would be a good place for the library, but if you want to host it under your own organization that's fine as well. If you decide to use MarketSquare, @emanlove can answer any questions and help with it.

**Comment by macpijan** [comment]
@pekkaklarck @emanlove MarketSquare sounds good to me. Can you please create a repository there. We can start moving the library?

**Comment by SergiiDmytruk** [comment]
So the library is to be moved to https://github.com/MarketSquare/robotframework-telnet-library by me.
I have a couple of questions on that, mainly about handling of tests.

#### Tests

The framework uses runner and tests that run tests through the runner.
Is that necessary to handle errors/exceptions from the library?
If so, is that because there used to be no `TRY/EXCEPT` syntax?

Looks like at least stripped-down version of `run.py` and `robot.xsd` will be necessary in standalone version of Telnet library, which shouldn't be a big deal unless this can result in a fragile setup (`run.py` depends on Robot's modules and not sure XML schema won't change).  So maybe using `TRY/EXCEPT` syntax should be the final goal here?

#### Eventual removal of the standard library

There are libdoc tests which seem to be using Telnet library as their input:
 * `atest/robot/libdoc/console_viewer.robot`
 * `atest/robot/libdoc/python_library.robot`

Should those be rewritten to use something else or maybe a copy of Telnet library could just be stored as test data?

The library is mentioned in several places including in examples.  Places which just reference it could link to its new location or just drop the mention, but Robot usage examples probably should switch to some other standard library,
right?

`AUTHORS.rst` has several mentions, but given that it's not being updated, I guess nothing needs to be done about it.

**Comment by emanlove** [comment]
Regarding the test runner I think it is there more as a convenience and tool to simplify development and not for any technical/execution reason. With such a large in depth and breadth software package I think the test runner is there to turn a complicated task into a simple singular command line to "run tests" and it takes care of everything for the developer. 

Here I think you could not include it initially if you have in place instructions on how to run the telnet specific tests. I see within a quick review of the tests there is a need for a local telnet server. So in this case a runner would set that all up for you. But for now one could just manually do all the steps. This is a good way to figure out hat is a pain to do and how does one simplify it (ie add a specific runner) that makes your life easier as maintainer of that library.

**Comment by SergiiDmytruk** [comment]
@emanlove, I probably wasn't clear enough given that there are 2 runners:
1. `atest/run.py`
2. `src/robot/run.py`

Your comment applies to the first one which is in fact a convenience, but it's the second one which I was wondering about.  

I now see that `robot` executable is basically an invocation of this second runner, it looked like it could be some internal detail at first.

XML schema might actually be validating XML output of Robot Framework as part of testing before feeding it back to one of framework's modules, meaning that a standalone library can drop the validation and assume RF is capable of understanding its own output.  So there should be no need to rewrite tests to get rid of nested runs of the framework, at least for now.

**Comment by pekkaklarck** [comment]
On holiday and on mobile so only very quick comments:
- The new library should probably have a totally new test setup. The old one uses Robot's normal `atest` infrastructure, but I don't think it has enough benefits in this case to warrant the extra complexity.
- There exists a separate `statuschecker` tool that can validate that keywords pass/fail as expected and emit correct log messages. You can see does it match your needs.
- Failures can also be tested easily with TRY/EXCEPT and the old Run Keyword And Expect Error works as well. The aforementioned `statuschecker` has a benefit that it allows validating log messages.
- You may want to check how e.g. SeleniumLibrary and Browser tests infra looks like.
- `src/robot/run.py` is part of Robot internals and not relevant to Telnet.

**Comment by MAPJe71-Delem** [comment]
FYI: We (Delem) are also users of the Telnet library, and would like to keep using it.

**Comment by chriskloe** [comment]
I am currently updating from a python 3.8 with robotframework 3 installation (I was using the RED plugin of ecliupse and thus was stuck to that version) to the current version and stumbled upon that telnet point as I am using telnet.
Is the new library available already or do I have to use an older python for a while?

**Comment by chriskloe** [comment]
Ok, thanks for the information! I am currently proceeding with Python 3.12 like this:
pipx reinstall-all --python3.12

**Comment by MarvKler** [comment]
Hi @macpijan are there any further steps planned for the development of the external telnet library for robotframework ? When can we expect to release a first version & what about the support for python 3.13 & higher versions ?

**Comment by pekkaklarck** [comment]
We have just made a decision to officially deprecate Telnet in RF 7.5 (#5786) and remove it in RF 8.0.

A precondition for the removal is that an external project with the same functionality exists. Unfortunately the project that was discussed above didn't really get started. Is someone else interested to create an external TelnetLibrary project and, hopefully, to maintain it also in the future? I'm happy to help with getting the project started and making the first release.

**Comment by wiktormowinski** [comment]
@pekkaklarck 
Hi,
As a folloup to [comment](https://github.com/robotframework/robotframework/issues/3537#issuecomment-2625038085) we're already working on a `TelnetLibrary` with the same functionality as the built-in one, and plan on submitting it as an external project in about 2 weeks.
